### PR TITLE
[CASH] Implement sensors calibration from MiscTA parameters

### PR DESCRIPTION
**Pull request by kholk**
This commit implements reading the sensors calibration from
their specific MiscTA units and stores them to a file in
the /data/misc/cashsvr/ directory, to avoid opening the TA
at every boot, which can be painfully slow.

This currently reads the parameters for both RGBCIR and
ToF sensors (currently, TCS3490 and STM VL53L0X), but will
only use the parameters for the ToF sensor - for now.

The parameters are sent to the ToF sensor currently via
sysfs are:
- Number of SPADs (type is fixed to not-aperture)
- Ranging offset calibration, expressed in micrometers

These parameters are required because, obviously, the
sensor sits behind a glass window.

### CODE REVIEW

File: cashsvr_input_miscta_params.c [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+		return 0;
+	}
+
+	fd = open(CASHSERVER_CALDATA_FILE, O_WRONLY | O_CREAT, 0664);
+	if (fd < 0) {
+		ALOGE("FATAL: Cannot open/create %s", CASHSERVER_CALDATA_FILE);
+		return -1;
+	} else if (set_perm) {
+		/* Set owner on the file, if it has been just created */
+		chown(CASHSERVER_CALDATA_FILE, AID_INPUT, AID_SYSTEM);
```
Comment (ix5) [inline cashsvr_input_miscta_params.c]: Best to use `getpwnam` like the rest of cash is doing, using `AID_*` is deprecated.

**EDIT:** Also `system:input` instead of `input:system` like the rest, please.

File: cashsvr_input_tof.c [hunk 2]
```diff
@@ -169,11 +175,87 @@ static int cash_tof_sys_init(bool high_accuracy, int devno, int plen)
...
+	if (spad_calib_path == NULL) {
+		free(buf);
+		goto calib_err;
+	}
+
+	snprintf(spad_calib_path, plen + LEN_REF_SPADS,
+		"%s%d/set_ref_spads", sysfs_input_str, devno);
+
+	if (chown(spad_calib_path, uid, gid) == -1) {
+		ALOGD("Cannot chown %s/set_ref_spads", spad_calib_path);
```
Comment (ix5) [inline cashsvr_input_tof.c]: strerror(errno) would be nice here

File: cashsvr_input_tof.c [hunk 3]
```diff
@@ -169,11 +175,87 @@ static int cash_tof_sys_init(bool high_accuracy, int devno, int plen)
...
+	if (um_offset_path == NULL) {
+		free(buf);
+		goto calib_err;
+	}
+
+	snprintf(um_offset_path, plen + LEN_REF_SPADS,
+		"%s%d/set_um_offset", sysfs_input_str, devno);
+
+	if (chown(um_offset_path, uid, gid) == -1) {
+		ALOGD("Cannot chown %s/set_um_offset", um_offset_path);
```
Comment (ix5) [inline cashsvr_input_tof.c]: same with strerror here

File: cashsvr_input_miscta_params.c [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+	if (stat(CASHSERVER_CALDATA_FILE, &st) == -1) {
+		set_perm = true;
+	} else if (!force) {
+		/* If file exists and not forcing re-read, just exit. */
+		return 0;
+	}
+
+	fd = open(CASHSERVER_CALDATA_FILE, O_WRONLY | O_CREAT, 0664);
+	if (fd < 0) {
+		ALOGE("FATAL: Cannot open/create %s", CASHSERVER_CALDATA_FILE);
```
Comment (ix5) [inline cashsvr_input_miscta_params.c:133]: same with strerror here

File: cash_private.h [hunk 5]
```diff
@@ -27,6 +27,16 @@
 #define CASHSERVER_TOF_CONF_FILE	"/vendor/etc/tof_focus_calibration.xml"
 #define CASHSERVER_RGBC_CONF_FILE	"/vendor/etc/cash_expcol_calibration.xml"
 
+#define CASHSERVER_DATASTORE_DIR	"/data/vendor/misc/cashsvr/"
```
Comment (ix5) [inline cash_private.h]: `/data/vendor/cashsvr` makes more sense now that I see the structure. `/data/misc/cashsvr` would have made sense under `/data/misc`, but nothing uses `/data/vendor/misc/*`
Either way, you need to add a `mkdir` in the cashsvr.rc file as well.

File: cashsvr_input_miscta_params.c [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+	return rc;
+}
+
+int cash_miscta_init_params(struct cash_tamisc_calib_params *conf)
+{
+	struct stat st = {0};
+	int ret, fd;
+
+	/* Create folder, if it doesn't exist */
```
Comment (ix5) [inline cashsvr_input_miscta_params.c]: Let's leave that to the cashsvr.rc init file, because... let's just say selinux isn't happy.

File: cashsvr_input_tof.c (hunk 2 above)
Comment (kholk) [reply cashsvr_input_tof.c]: chown return zero for success or -1 for error.

File: cashsvr_input_miscta_params.c (hunk 4 above)
Comment (kholk) [reply cashsvr_input_miscta_params.c:133]: open returns a new file descriptor or -1 for error.

**Comment by kholk** [comment]
The commit was amended and the branch has been forcepushed.

Now it should be good.
Please test!

File: cashsvr_input_miscta_params.c [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+	miscta_lnk.ta_read = dlsym(miscta_lnk.ta_handle, "ta_read");
+	if (!miscta_lnk.ta_read) {
+		ALOGE("%s: DLSYM failed for ta_read", __func__);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int cash_miscta_read_unit(uint32_t id, void *data_out)
```
Comment (MarijnS95) [inline cashsvr_input_miscta_params.c:90]: Consider passing the size of `data_out` as well. While we are not expecting this to change, ever, it's better to program some resilience against accidental writes out of the memory region.
This'll help bringing up newer sensors that may or may not have different sizes/parameters (again, if ever).

File: cashsvr_input_miscta_params.c [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+	} else if (!force) {
+		/* If file exists and not forcing re-read, just exit. */
+		return 0;
+	}
+
+	fd = open(CASHSERVER_CALDATA_FILE, O_WRONLY | O_CREAT, 0664);
+	if (fd < 0) {
+		ALOGE("FATAL: Cannot open/create %s", CASHSERVER_CALDATA_FILE);
+		return -1;
+	} else if (set_perm) {
```
Comment (MarijnS95) [inline cashsvr_input_miscta_params.c]: Can leave out the `else` since the above block returns. Adds to readability.

File: cashsvr_input_miscta_params.c (hunk 6 above)
Comment (MarijnS95) [reply cashsvr_input_miscta_params.c]: Update the comment, since creation was removed :smile:

File: vendor/etc/init/cashsvr.rc [hunk 9]
```diff
@@ -1,6 +1,7 @@
 on post-fs-data
-    # create socket for cashsvr
+    # create socket and data directory for cashsvr
     mkdir /dev/socket/cashsvr 0755 system system
+    mkdir /data/vendor/cashsvr 0755 system system
```
Comment (ix5) [inline vendor/etc/init/cashsvr.rc]: If this dir is owned by `system:system`, root will have to use `dac_override` to create a file in it
So, let's use `root root`

File: cashsvr.c [hunk 10]
```diff
@@ -493,12 +494,18 @@ int cashsvr_configure(void)
 	cash_conf.rgbc_polyreg_extra = 0;
 	cash_conf.disable_rgbc = 0;
 
+        /*
+         * Retrieve the calibration data either from MiscTA
+         * or from CASH's calibration file
+         */
+        cash_miscta_init_params(&calib_params);
```
Comment (MarijnS95) [inline cashsvr.c]: Mixing tabs and spaces here.

File: cashsvr_input_miscta_params.c [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+	if (ret)
+		return ret;
+
+	fd = open(CASHSERVER_CALDATA_FILE, O_RDONLY);
+	if (fd < 0)
+		return -1;
+
+	read(fd, conf, sizeof(struct cash_tamisc_calib_params));
+	close(fd);
+
```
Comment (MarijnS95) [inline cashsvr_input_miscta_params.c:223]: Can we print the configuration here? That'll help in verifying, as well as showing the actual meaning of these values (caps sound like a black box - do they define a calibration curve?).

File: cashsvr_input_tof.c [hunk 12]
```diff
@@ -151,29 +137,86 @@ static int cash_tof_sys_init(bool high_accuracy, int devno, int plen)
...
+		free(buf);
+		goto calib_err_mem;
+	}
+
+	snprintf(spad_calib_path, plen + LEN_REF_SPADS,
+		"%s%d/set_ref_spads", sysfs_input_str, devno);
+
+	rc = cash_set_permissions(spad_calib_path, "system", "input");
+	if (rc == -1) {
+		free(spad_calib_path);
```
Comment (MarijnS95) [inline cashsvr_input_tof.c:176]: `buf` not freed here.

File: cashsvr_input_tof.c [hunk 13]
```diff
@@ -151,29 +137,86 @@ static int cash_tof_sys_init(bool high_accuracy, int devno, int plen)
...
+		if (rc == -1) {
 			ALOGD("Cannot chown %s/enable_ps_sensor", sns_mode_path);
-			return 1;
-		}
-
-		fd = open(sns_mode_path, O_WRONLY);
-		if (fd < 0) {
-			ALOGD("Cannot open %s", sns_mode_path);
-			return 1;
+			free(cash_tof_enable_path);
```
Comment (MarijnS95) [inline cashsvr_input_tof.c:141]: Consider setting `cash_tof_enable_path` to `NULL`, as it's checked against. Same for the `free` at line 133.

File: cashsvr_input_tof.c [hunk 14]
```diff
@@ -151,29 +137,86 @@ static int cash_tof_sys_init(bool high_accuracy, int devno, int plen)
...
+			free(sns_mode_path);
+			return rc;
 		}
 
-		rc = write(fd, VL53L0_HIGH_ACCURACY, 1);
+		rc = cash_set_parameter(sns_mode_path, VL53L0_HIGH_ACCURACY, 1);
```
Comment (MarijnS95) [inline cashsvr_input_tof.c]: `sns_mode_path` Not used after this line. Consider freeing it.

File: cash_input_common.c [hunk 15]
```diff
@@ -71,3 +69,43 @@ int cash_set_parameter(char* path, char* value, int value_len) {
...
+/*
+ * cash_set_permissions - Common utility function to set file permissions
+ *
+ * \param fpath - Path of the file or directory
+ * \param str_uid - Username string
+ * \param str_gid - Group name string
+ *
+ * \return Returns zero for success or -1 for error.
+ */
+int cash_set_permissions(char* fpath, char* str_uid, char* str_gid)
```
Comment (MarijnS95) [inline cash_input_common.c:82]: Shouldn't the arguments be `const char *`?

File: cashsvr_input_tof.c [hunk 16]
```diff
@@ -151,29 +137,86 @@ static int cash_tof_sys_init(bool high_accuracy, int devno, int plen)
...
+		free(buf);
+		goto calib_err_mem;
+	}
+
+	snprintf(um_offset_path, plen + LEN_REF_SPADS,
+		"%s%d/set_um_offset", sysfs_input_str, devno);
+
+	rc = cash_set_permissions(um_offset_path, "system", "input");
+	if (rc == -1) {
+		free(um_offset_path);
```
Comment (MarijnS95) [inline cashsvr_input_tof.c:196]: `buf` not freed here.

File: cashsvr_input_tof.c [hunk 17]
```diff
@@ -151,29 +137,86 @@ static int cash_tof_sys_init(bool high_accuracy, int devno, int plen)
 		snprintf(sns_mode_path, plen + LEN_MODE,
 			"%s%d/set_use_case", sysfs_input_str, devno);
 
-		if (chown(sns_mode_path, uid, gid) == -1) {
+		rc = cash_set_permissions(sns_mode_path, "system", "input");
+		if (rc == -1) {
 			ALOGD("Cannot chown %s/enable_ps_sensor", sns_mode_path);
```
Comment (MarijnS95) [inline cashsvr_input_tof.c]: Just `%s`.

File: cashsvr_input_tof.c (hunk 13 above)
Comment (MarijnS95) [reply cashsvr_input_tof.c:141]: Also, why is the global enable path freed in this error case, but not in the error cases below?

File: cashsvr_input_tof.c (hunk 14 above)
Comment (MarijnS95) [reply cashsvr_input_tof.c]: Consider `sizeof(VL53L0_HIGH_ACCURACY) - 1` for the size - prevents undefined behaviour when changing the define in the future.

**Review by MarijnS95 — COMMENTED** [review]
Please review the allocations and free's again.

Also, some "feature" requests to make this a little more future proof (even if we are not expecting this to change any time soon).

**Comment by djselbeck** [comment]
Lilac is missing the calibration xmls. Is there information available how to obtain or create these?

**Comment by jerpelea** [comment]
@kholk

**Comment by ix5** [comment]
> Lilac is missing the calibration xmls. Is there information available how to obtain or create these?

You can look to kagura for reference: https://github.com/sonyxperiadev/device-sony-kagura/blob/master/rootdir/vendor/etc/tof_focus_calibration.xml

But beware, this is probably not going to fit lilac's sensor.

**Comment by djselbeck** [comment]
> > Lilac is missing the calibration xmls. Is there information available how to obtain or create these?
> 
> You can look to kagura for reference: https://github.com/sonyxperiadev/device-sony-kagura/blob/master/rootdir/vendor/etc/tof_focus_calibration.xml
> 
> But beware, this is probably not going to fit lilac's sensor.

Poplar's calibration works quite good

**Comment by kholk** [comment]
Of course looking at a IMX300-related calibration for an IMX400 device would be totally stupid....

But yeah, I can create a guide on how to calibrate your actuator based on ToF distance parameters.
I will do it whenever I get some spare time to invest in that.

**Comment by kholk** [comment]
Forcepushed again to address the last critical issues.
I didn't integrate the proposed enhancements, like the logging nor constification of the function parameters for set_permissions.

Please check if the code works for you, I'm pretty sure it works fine, but I don't know if I've messed the commit somehow when ping-pong'ing between a machine and another.

If the code works, it can be merged.
For any other enhancements, please open a PR after this is merged.

Thanks!

**Comment by russel5** [comment]
> But yeah, I can create a guide on how to calibrate your actuator based on ToF distance parameters.
I will do it whenever I get some spare time to invest in that.

Bro, you tell that first time 7 month ago and we still not have proper calibration xml file on lilac

File: cashsvr_input_tof.c [hunk 18]
```diff
@@ -101,79 +102,120 @@ int cash_tof_enable(bool enable)
...
+		rc = cash_set_permissions(sns_mode_path, "system", "input");
+		if (rc == -1) {
+			free(cash_tof_enable_path);
+			free(sns_mode_path);
+			cash_tof_enable_path = NULL;
+			return rc;
 		}
 
-		rc = write(fd, VL53L0_HIGH_ACCURACY, 1);
+		rc = cash_set_parameter(sns_mode_path, VL53L0_HIGH_ACCURACY,
```
Comment (MarijnS95) [inline cashsvr_input_tof.c:147]: `sns_mode_path` not used beyond this line, should it be freed?

**Review by MarijnS95 — APPROVED** [review]
Thanks for the fixes, looks much better! One allocation is not freed, other than that it looks good now :)

**Comment by ix5** [comment]
**Edit:** With the `chown()` removed in the calib data init function and checking for `rc < 0` when calling `cash_set_parameter`, no further errors.

**EDIT 2:** With some more logging and a fresh reboot, now I get:
```
04-29 07:43:11.265  1724  1724 I CASH    : Initializing Camera Augmented Sensing Helper Server...
04-29 07:43:11.286  1724  1724 E CASH_INPUT_COMMON: getpwnam(str_uid=system) uid=1000 gid=1000
04-29 07:43:11.286  1724  1724 E CASH_INPUT_COMMON: getpwnam(str_gid=input) uid=1004 gid=1004
04-29 07:43:11.286  1724  1724 E CASH_INPUT_COMMON: getpwnam(str_uid=system) uid=1000 gid=1000
04-29 07:43:11.286  1724  1724 E CASH_INPUT_COMMON: getpwnam(str_gid=input) uid=1004 gid=1004
04-29 07:43:11.287  1724  1724 W CASH_TOF_INPUT: ERROR! Cannot set ToF High Accuracy mode! No such file or directory
04-29 07:43:11.287  1724  1724 E CASH_INPUT_COMMON: getpwnam(str_uid=system) uid=1000 gid=1000
04-29 07:43:11.287  1724  1724 E CASH_INPUT_COMMON: getpwnam(str_gid=input) uid=1004 gid=1004
04-29 07:43:11.287  1724  1724 E CASH_TOF_INPUT: ERROR! Cannot set Reference SPADs! No such file or directory
04-29 07:43:11.288  1724  1724 E CASH_INPUT_COMMON: getpwnam(str_uid=system) uid=1000 gid=1000
04-29 07:43:11.288  1724  1724 E CASH_INPUT_COMMON: getpwnam(str_gid=input) uid=1004 gid=1004
04-29 07:43:11.288  1724  1724 E CASH_TOF_INPUT: ERROR! Cannot set micrometer offset! No such file or directory
```

Files have incorrect permissions:
```
-rw-rw-r-- 1 input input u:object_r:sysfs_tof_sensor:s0 4096 1970-04-29 07:42 enable_ps_sensor
```
Already checked, all the calls to `getpwnam` and chown seem correct...

---

**Original post**: Getting this in <strike>enforcing</strike> permissive:
```
CASH    : Initializing Camera Augmented Sensing Helper Server...
auditd  : type=1400 audit(0.0:7): avc: denied { dac_override } for comm="cashsvr" \
  capability=1 scontext=u:r:cashsvr:s0 tcontext=u:r:cashsvr:s0 tclass=capability permissive=1
cashsvr : type=1400 audit(0.0:7): avc: denied { dac_override } for capability=1 \
  scontext=u:r:cashsvr:s0 tcontext=u:r:cashsvr:s0 tclass=capability permissive=1
CASH_TOF_INPUT: ERROR! Cannot set ToF High Accuracy mode!
CASH_TOF_INPUT: ERROR! Cannot set Reference SPADs!
CASH_TOF_INPUT: ERROR! Cannot set micrometer offset!
CASH    : Table x:103.00  y:10000000.00
```

File: cash_input_common.c [hunk 19]
```diff
@@ -17,13 +17,11 @@
...
  */
 
-#include <sys/poll.h>
-#include <sys/epoll.h>
 #include <errno.h>
 #include <stdbool.h>
 #include <pthread.h>
 #include <unistd.h>
-
+#include <pwd.h>
```
Comment (ix5) [inline cash_input_common.c:24]: Missing a `LOG_TAG`

File: cashsvr_input_tof.c (hunk 18 above)
Comment (ix5) [reply cashsvr_input_tof.c:147]: Also, `cash_set_parameter` returns 0 on success, so check for `rc < 0` instead of `rc < 1`.

File: cashsvr_input_tof.c [hunk 20]
```diff
@@ -101,79 +102,120 @@ int cash_tof_enable(bool enable)
...
+	rc = cash_set_permissions(spad_calib_path, "system", "input");
+	if (rc == -1) {
+		free(spad_calib_path);
+		goto calib_err;
+	}
+
+	cnt = snprintf(buf, 5, "%u", calib_params->tof_spad_num);
+
+	rc = cash_set_parameter(spad_calib_path, buf, (cnt * sizeof(char)));
+	if (rc < 1)
```
Comment (ix5) [inline cashsvr_input_tof.c]: Same here, check for `rc < 0` instead

File: cashsvr_input_tof.c [hunk 21]
```diff
@@ -101,79 +102,120 @@ int cash_tof_enable(bool enable)
...
+	rc = cash_set_permissions(um_offset_path, "system", "input");
+	if (rc == -1) {
+		free(um_offset_path);
+		goto calib_err;
+	}
+
+	cnt = snprintf(buf, 5, "%u", calib_params->tof_um_offset);
+
+	rc = cash_set_parameter(um_offset_path, buf, (cnt * sizeof(char)));
+	if (rc < 1)
```
Comment (ix5) [inline cashsvr_input_tof.c]: Same here, check for `rc < 0` instead

**Review by ix5 — CHANGES_REQUESTED** [review]

File: cash_input_common.c (hunk 19 above)
Comment (kholk) [reply cash_input_common.c:24]: That's on purpose

**Comment by kholk** [comment]
Forcepushed to solve the error checking stupidities and the missing free

File: cash_input_common.c (hunk 19 above)
Comment (ix5) [reply cash_input_common.c:24]: Errors from cash_input_comon.c don't show up when you grep logcat for `cash`. Setting a LOG_TAG here fixes that.

Comment (kholk) [reply cash_input_common.c:24]: The cash_input_common contains common code that is shared between all "modules" in CASH.
If there is any error coming from a cash_input_common function, unless this is failing on _all CASH modules_, then it's relative to a specific one.

For example, if I make a wrong call to a function in cash_input_common from cash_input_tof, that's a cash_input_tof error, not a common one, hence I want to see that the error comes from the ToF module.
Being it common, it could be happening on the RGBCIR module, as well, so if I see "CASH_INPUT_COMMON: Cannot allocate", I don't know if it comes from the ToF or TGBCIR module.

That's why I don't put a LOG_TAG here.

Comment (ix5) [reply cash_input_common.c:24]: Makes sense, thx

Comment (MarijnS95) [reply cash_input_common.c:24]: I disagree. No `LOG_TAG` defined means that any logging printed from this compilation unit has an empty tag. As @ix5 correctly mentions, using a simple grep for `-i cash` now has missing lines while debugging.
If you want to see which module the error comes from, you should print an error in the respective module when a common function returns an error code. Having the logging in common itself aids in pinpointing the exact reason for that case (without deciphering the error code to print something useful in the tof/rgbcir module).

tl;dr: Either remove the common ALOG* calls altogether, or associate them with `cash`.

**Comment by kholk** [comment]
This is working. Can we merge this?
